swarm/storage/localstore.go

```diff
@@ -195,31 +195,42 @@ func (ls *LocalStore) Close() {
 
 // Migrate checks the datastore schema vs the runtime schema, and runs migrations if they don't match
 func (ls *LocalStore) Migrate() error {
-	schema, err := ls.DbStore.GetSchema()
+	actualDbSchema, err := ls.DbStore.GetSchema()
 	if err != nil {
 		log.Error(err.Error())
 		return err
 	}
 
-	log.Debug("found schema", "schema", schema, "runtime-schema", CurrentDbSchema)
-	if schema != CurrentDbSchema {
-		// run migrations
-
-		if schema == "" {
-			log.Debug("running migrations for", "schema", schema, "runtime-schema", CurrentDbSchema)
+	log.Debug("DB schema", "current", actualDbSchema, "desired", DesiredDbSchema)
+	if actualDbSchema == DesiredDbSchema {
+		return nil
+	}
 
-			// delete chunks that are not valid, i.e. chunks that do not pass any of the ls.Validators
-			ls.DbStore.Cleanup(func(c *chunk) bool {
-				return !ls.isValid(c)
-			})
+	if actualDbSchema == DbSchemaNone {
+		ls.migrateFromNoneToPurity()
+		actualDbSchema = DbSchemaPurity
+	}
 
-			err := ls.DbStore.PutSchema(DbSchemaPurity)
-			if err != nil {
-				log.Error(err.Error())
-				return err
-			}
-		}
+	if actualDbSchema == DbSchemaPurity {
+		ls.migrateFromPurityToHalloween()
+		actualDbSchema = DbSchemaHalloween
 	}
 
+	if err := ls.DbStore.PutSchema(actualDbSchema); err != nil {
+		log.Error(err.Error())
+		return err
+	}
 	return nil
 }
+
+func (ls *LocalStore) migrateFromNoneToPurity() {
+	// delete chunks that are not valid, i.e. chunks that do not pass
+	// any of the ls.Validators
+	ls.DbStore.Cleanup(func(c *chunk) bool {
+		return !ls.isValid(c)
+	})
+}
+
+func (ls *LocalStore) migrateFromPurityToHalloween() {
+	ls.DbStore.CleanGCIndex()
```

Comment:
i'd rather see the migration code extracted from the main codebase, even better - put it here. we shouldn't add so much code to our usual bread and butter files - there's enough complexity there as it is and there's no reason to plant a one-time executing function there forever

Reply:
I totally disagree. 
Why postponing this? Is this so complex to do?
`v`+`dd`+`P` is not too much to ask

Reply:
I now realize I misunderstood what you meant here. If it's just moving the block of code into a separate `.go` file of course that's no biggie.

Reply:
So we don't have a clear and thought through solution for this. The db is fubar and swarm is thus useless until this fix is applied. The structuring of migration is a separate problem (it was originally even a separate PR when it was introduced), so let's defer this to a separate PR.